### PR TITLE
GUI: Allow text selection

### PR DESCRIPTION
**Pull request by srikeerthis**
This PR adds the enhancement requested at
https://bugs.scummvm.org/ticket/6948

The key combination Shift+Left Arrow will highlight and select text towards left of the caret.

Two methods have been added:
- initHighlight() - resets highlight.
- getHighlightOffset() - used for scrolling text animation as text is selected.

Files changed:
- editable.h: declared variables and methods.
- edittext.cpp: Code segment that prevents highlight from being drawn on any position where mouse is clicked.
- editable.cpp:
  - Select and highlight text towards left.
  - Copy selected text using Ctrl-C.
  - Remove selected characters on Backspace/Delete key.
  - Replace selected text with a character.
  - Disable and re-initialize highlight on press of Left Arrow/Right Arrow/Home/End keys

### CODE REVIEW

File: gui/widgets/editable.cpp [hunk 1]
```diff
@@ -20,11 +20,11 @@
  *
  */
 
+#include "gui/widgets/editable.h"
```
Comment (sev-) [inline gui/widgets/editable.cpp:23]: These changes must be reverted as they're just reordering of includes

File: gui/widgets/editable.cpp [hunk 2]
```diff
@@ -129,49 +138,86 @@ bool EditableWidget::handleKeyDown(Common::KeyState state) {
 	switch (state.keycode) {
 	case Common::KEYCODE_RETURN:
 	case Common::KEYCODE_KP_ENTER:
+		initHighlight();
```
Comment (sev-) [inline gui/widgets/editable.cpp:141]: the name of the method is misleading. "init" normally means start. So, better name it `clearHighLight()`

File: gui/widgets/editable.cpp [hunk 3]
```diff
@@ -181,23 +227,31 @@ bool EditableWidget::handleKeyDown(Common::KeyState state) {
...
 		forcecaret = true;
 		dirty = true;
 		break;
 
 	case Common::KEYCODE_UP:
 	case Common::KEYCODE_HOME:
+		initHighlight();
 		// Move caret to start
 		setCaretPos(0);
+		sendCommand(_cmd, 0);
```
Comment (sev-) [inline gui/widgets/editable.cpp:244]: What is the purpose of this call?

File: gui/widgets/editable.cpp [hunk 4]
```diff
@@ -262,8 +316,25 @@ bool EditableWidget::handleKeyDown(Common::KeyState state) {
 }
 
 void EditableWidget::defaultKeyDownHandler(Common::KeyState &state, bool &dirty, bool &forcecaret, bool &handled) {
-	if (state.ascii < 256 && tryInsertChar((byte)state.ascii, _caretPos)) {
+	//  if highlighted, replace selected text with inserted character
+	if (_highlightVisible) {
+		if (state.ascii < 256 && tryInsertChar((byte)state.ascii, _caretPos)) {
+			for (uint32 i = _highlightPos + _highlightCharacterCount; i > (uint32)_highlightPos; i--) {
+				_editString.deleteChar(i - 1);
```
Comment (sev-) [inline gui/widgets/editable.cpp:323]: please use `Common::String::erase(uint32 pos, uint32 len)` instead of this loop.

File: gui/widgets/editable.cpp [hunk 5]
```diff
@@ -341,8 +441,8 @@ void EditableWidget::drawCaret(bool erase) {
 		width = MIN(editRect.width() - caretOffset, width);
 		if (width > 0) {
 			g_gui.theme()->drawText(Common::Rect(x, y, x + width, y + editRect.height()), character,
-			                        _state, _drawAlign, _inversion, 0, false, _font,
```
Comment (sev-) [inline gui/widgets/editable.cpp:344]: This whitespace change should be reverted, or put into a separate commit

File: gui/widgets/editable.cpp [hunk 6]
```diff
@@ -389,4 +503,19 @@ void EditableWidget::makeCaretVisible() {
...
 }
 
+void EditableWidget::initHighlight() {
+	// disable highlight
+	_highlightVisible = false;
+	// Empty the temporary string of previous values
+	_highlightString.clear();
+	// insert blank characters
+	for (int i = 0; i < (int)_caretPos; i++)
+		_highlightString.insertChar(' ', i);
```
Comment (sev-) [inline gui/widgets/editable.cpp:513]: What is the logic behind this?

**Review by sev- — CHANGES_REQUESTED** [review]
I reviewed the code and put my comments and questions.

A quick test showed the following:

- [ ] The highlight interferes with the tooltip. If I move the mouse over e.g. game id and start highlighting, the highlight will blink when the tooltip is shown
- [ ] On initial launch and starting highlighting right away, it leads to the visual glitch when whole line becomes green, with flickering white spot where cursor is supposed to be, and doubled input line. I'll attach a screenshot
- [ ] When selecting, clicking "shift-right" clears the whole selection and moves cursor right. The logical thing would be to decrease the highlight size by one
- [ ] If the line is wider than the input (scrollPos is non-zero), highlighting on the left edge leads to the whole line being highlighted
- [ ] When in the middle of a line, Shift-Right does not start the highlighting

**Comment by sev-** [comment]
Screenshot 2021-04-19 at 13 36 29

Screenshot of glitched highlight. The input string is "kyra1"

File: gui/widgets/editable.cpp (hunk 6 above)
Comment (srikeerthis) [reply gui/widgets/editable.cpp:513]: I have tried using append() method to add characters, but had difficulty in reversing the string and displaying the same. So the logic thought was to have an empty temporary string equal to the size of the caret position and add characters from the end of the temporary string.

**Comment by sev-** [comment]
Closing this due to inactivity, lack of handling of the edge cases.
